src/macros.rs

```diff
@@ -0,0 +1,69 @@
+/// Used for creating container type that self-reference owned data,
+/// see also <https://morestina.net/blog/1868/self-referential-types-for-fun-and-profit>
+///
+/// # Example
+///
+/// ```ignore
+/// struct Data();
+/// struct SomeRef<'a>(&'a Data);
+///
+/// mod inner {
+///     impl_self_ref!(Container, SomeRef<'static>, SomeRef<'this>);
+/// }
+/// use inner::Container;
+///
+/// let container = Container::new(Data(), |data| SomeRef(data));
+/// let some_ref = container.as_ref();
+/// ```
+macro_rules! impl_self_ref {
+    ($SelfRef:ident, $RefStatic:ty, $RefThis:ty) => {
+        /// # Safety invariant
+        ///
+        /// `data` must never have a mutable reference taken, nor be modified during the lifetime of
+        /// this struct
+        pub struct $SelfRef<T> {
+            /// `data_ref` could self-referencing `data`
+            data_ref: $RefStatic,
```

Comment:
This field needs to be wrapped in `maybe_dangling::MaybeDangling` or `core::mem::MaybeUninit` to prevent the compiler from potentially assuming any references it contains will be valid till the end of a function when `$SelfRef<T>` is passed as a function parameter (see the issue I linked above). Any such references will only be valid until this value is dropped.